packages/fluentui/react-northstar/src/components/Tree/hooks/useTreeSelectState.tsx

```diff
@@ -0,0 +1,136 @@
+import * as React from 'react';
+import { useAutoControlled } from '@fluentui/react-bindings';
+import * as _ from 'lodash';
+import { TreeProps } from '../Tree';
+import { useStableProps } from './useStableProps';
+import { FlatTree } from './flattenTree';
+import { findIndex, removeItemAtIndex } from './utils';
+
+export interface TreeSelectState {
+  selectedItemIds: string[];
+  flatTree: FlatTree;
+  toggleSelect: (ids: string[], e: React.SyntheticEvent) => void;
+}
+
+export function useTreeSelectState(
+  props: Pick<TreeProps, 'defaultSelectedItemIds' | 'selectedItemIds' | 'items'>,
+  flatTree: FlatTree,
+): TreeSelectState {
+  // selectedItemIds is only valid for leaf nodes.
+  // For non-leaf nodes, their 'selected' states are defered from all their descendents
+  const [selectedItemIds, setSelectedItemIdsState] = useAutoControlled<string[]>({
+    defaultValue: props.defaultSelectedItemIds,
+    value: props.selectedItemIds,
+    initialValue: [],
+  });
+
+  // compute a new flattened tree based on selectedItemIds state
+  const updatedFlatTree = React.useMemo(() => {
+    // build updatedFlatTree based on flatTree, with all tree item having selected false.
+    // then updates the 'selected' prop of leaves based on selectedItemIds
+    const updatedFlatTree = {};
+    const rootsIds = [];
+    Object.keys(flatTree).forEach(id => {
+      updatedFlatTree[id] = {
+        ...flatTree[id],
+        selected: false,
+      };
+      if (!updatedFlatTree[id].parent) {
+        rootsIds.push(id);
+      }
+    });
+    selectedItemIds.forEach(id => {
+      if (!updatedFlatTree[id].hasSubtree) {
+        updatedFlatTree[id].selected = true;
+      }
+    });
+
+    // traverse all tree nodes in updatedFlatTree for once,
+    // to calculate the selection state of the parent nodes based on leaf nodes
+    const traverseTree = nodes => {
```

Comment:
consider better naming, a function called `traverseTree` that returns a boolean or 'indeterminate' doesn't really tell me what it does

Reply:
do you even use the return values ?

Reply:
Yes it is used in line 56.

This function may get changed later. Shift hasn't reviewed this part yet, so I'm not sure if my implementation here is good enough.

This `traverseTree` is used for calculating the select state of a tree. 
The parent tree node can have selected/not-selected/indeterminate state, based on if its children is fully-selected/not-selected/half-selected.
So `traverseTree`, is doing a depth first traverse down the tree, til it reaches leaf nodes, then gather the selection state of all leaf children, and then pass this information back to parent. And parent to its parent, and so on.

It's really hard for me to came up a name to make sense here.... 🙈

Reply:
alright, that's fine, let's resolve this comment for now if you'll revisit this